tools/archives/pyvo_integration/astronomical_archives.py

```diff
@@ -0,0 +1,300 @@
+            'service_type': ''
+        }
+
+        if self.keyword:
+            parameters['keywords'] = self.keyword
+
+        if Waveband.is_waveband_supported(self.waveband):
+            parameters['waveband'] = Waveband.wavebands[self.waveband]
+
+        if Service.is_service_supported(self.service_type):
+            parameters['service_type'] = Service.services[self.service_type]
+        else:
+            parameters['service_type'] = Service.services['TAP']
+
+        return parameters
+
+
+class Registry:
+
+    def __init__(self):
+        pass
+
+    @staticmethod
+    def search_registries(rsp: RegistrySearchParameters, number_of_registries):
+
+        parameters = rsp.get_parameters()
+
+        keywords = parameters['keywords']
+        waveband = parameters['waveband']
+        service_type = parameters['service_type']
+
+        registry_list = []
+
+        if not waveband:
+            registry_list = registry.search(keywords=keywords, servicetype=service_type)
+        else:
+            registry_list = registry.search(keywords=keywords, waveband=waveband, servicetype=service_type)
+
+        if registry_list:
+            registry_list = Registry._get_registries_from_list(registry_list, 1)
+
+        return registry_list
+
+    @staticmethod
+    def _get_registries_from_list(registry_list, number_of_registries):
+
+        archive_list = []
+
+        for i, registry in enumerate(registry_list):
+            if i < number_of_registries:
+                archive = TapArchive(registry.standard_id, registry.res_title, registry.short_name, registry.access_url)
+                archive_list.append(archive)
+
+        return archive_list
+
+
+class TapQuery:
+
+    def __init__(self, query):
+        self.raw_query = query
+
+    def get_query(self):
+        return urllib.parse.unquote(self.raw_query).replace("+", " ")
+
+
+class FileHandler:
+
+    def __init__(self):
+        pass
+
+    @staticmethod
+    def download_file_from_url(file_url):
+        with request.urlopen(file_url) as response:
+            fits_file = response.read()
+
+        return fits_file
+
+    @staticmethod
+    def create_file_from_url_list(file_urls):
+        url_file = ''
+
+        for url in file_urls:
+            url_file + url + ' '
+
+        return url_file
+
+    @staticmethod
+    def write_file_to_output(file, output):
+        with open(output, "wb") as file_output:
+            file_output.write(file)
+
+
+if __name__ == "__main__":
+
+    output = sys.argv[1]
+
+    keyword = sys.argv[2]
+    waveband = sys.argv[3]
+    service_type = sys.argv[4]
+    adql_query = sys.argv[5]
+
+    file_type = sys.argv[6]
+    number_of_files = sys.argv[7]
+    tabular_output = sys.argv[8]
+
+    file_url = []
+
+    rsp = RegistrySearchParameters(keyword=keyword, waveband=waveband, service_type='TAP')
+
+    archive_list = Registry.search_registries(rsp, 1)
+
+    if archive_list:
+        archive_list[0].initialize()
```

Comment:
so you will just query fro the first archive? It's not what users wil want.

Did you not have a version were it was possible to select archives as input?

Reply:
The selection of specific archives rely on the possibility to dynamically create the tool interface which I wasn't able to do (I think there be might a way (as the upload tool loads a list of datatypes from calling some galaxy registry file) but I wasn't able to figure it out and also i'm pretty sure it would require adding some code to core galaxy which can't be done if it's a simple tool in the toolshed)

Another possibility would be to download every archive tap service url into a tool data file and populate a list from there which I think could work but it means the list should be regularly updated to account for changes in archives

Reply:
The second possibility is what I also suggested as a first implementation, let's do this, and then ask our Galaxy colleagues for advice.